File: ShibuyaTools/utils/Phrases.fs
module Phrases

let private isValid (phrase: string) =
    phrase.Length > 0 && not(Seq.exists (fun c -> int c < 20 || int c > 65532) phrase)

let private encoding = System.Text.UTF8Encoding(false)

let detect bytes = seq {
    let mutable f = false
    let mutable s = 0

    for i = 0 to Array.length bytes - 1 do
        match bytes.[i] with
        | 1uy when not f ->
            f <- true
            s <- i + 1
        | 1uy when f ->
            s <- i + 1
        | 2uy when f ->
            f <- false
            let phrase = encoding.GetString(bytes[s..i-1])
            if isValid phrase then yield (s, i, phrase)
        | _ -> ()
    }

let encode (s: string) = encoding.GetBytes(s)

File: ShibuyaTools.Resources.Wad/Phrases.cs
using System.Text;

namespace ShibuyaTools.Resources.Wad;

internal static class Phrases
{
    private static readonly Encoding Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);

    public static byte[] Encode(string s) => Encoding.GetBytes(s);

    public static IEnumerable<(int Start, int End, string Phrase)> Detect(byte[] bytes)
    {
        var f = false;
        var s = 0;

        for (var i = 0; i < bytes.Length; i++)
        {
            switch (bytes[i])
            {
                case 1 when !f:
                    f = true;
                    s = i + 1;
                    break;

                case 1 when f:
                    s = i + 1;
                    break;

                case 2 when f:
                    f = false;
                    var phrase = Encoding.GetString(bytes[s..i]);
                    if (IsValid(phrase)) yield return (s, i, phrase);
                    break;
            }
        }
    }

    private static bool IsValid(string phrase)
    {
        return phrase.Length > 0 && !phrase.Any(c => c < 20 || c > 65532);
    }
}
